src/charwise/builder.rs

```diff
@@ -294,136 +293,53 @@ impl CharwiseDoubleArrayAhoCorasickBuilder {
-            }
-            if i == self.block_len - 1 {
-                self.set_next(i, 0);
-            } else {
-                self.set_next(i, i + 1);
-            }
-        }
-        self.fix_state(ROOT_STATE_IDX);
-    }
-
-    #[inline(always)]
-    fn fix_state(&mut self, idx: u32) {
-        // DEAD_STATE_IDX-th element should be always free.
-        debug_assert_ne!(idx, DEAD_STATE_IDX);
-        debug_assert!(!self.is_fixed(idx));
-
-        let next = self.get_next(idx);
-        let prev = self.get_prev(idx);
-        self.set_next(prev, next);
-        self.set_prev(next, prev);
-        self.set_fixed(idx);
+            .resize(usize::try_from(self.block_len).unwrap(), State::default());
+        let mut helper = BuildHelper::new(self.block_len, self.num_free_blocks);
+        helper.push_block().unwrap();
+        helper.use_index(ROOT_STATE_IDX);
+        helper.use_index(DEAD_STATE_IDX);
+        helper
     }
 
-    // edges = [(char, next_id)] sorted by char
     #[inline(always)]
-    fn find_base(&self, edges: &[(u32, u32)]) -> u32 {
+    fn find_base(&self, edges: &[(u32, u32)], helper: &BuildHelper) -> u32 {
         debug_assert!(!edges.is_empty());
 
-        let mut idx = self.get_next(DEAD_STATE_IDX);
-        while idx != DEAD_STATE_IDX {
-            debug_assert!(!self.is_fixed(idx));
+        for idx in helper.vacant_iter() {
             let base = idx ^ edges[0].0;
-            if self.verify_base(base, edges) {
+            if self.verify_base(base, edges, helper) {
                 return base;
             }
-            idx = self.get_next(idx);
         }
         u32::try_from(self.states.len()).unwrap() ^ edges[0].0
     }
 
     #[inline(always)]
-    fn verify_base(&self, base: u32, edges: &[(u32, u32)]) -> bool {
+    fn verify_base(&self, base: u32, edges: &[(u32, u32)], helper: &BuildHelper) -> bool {
         for &(c, _) in edges {
             let idx = base ^ c;
-            if self.is_fixed(idx) {
+            if helper.is_used_index(idx) {
                 return false;
             }
         }
         true
     }
 
     #[inline(always)]
-    fn extend_array(&mut self) {
-        let old_len = u32::try_from(self.states.len()).unwrap();
-        let new_len = old_len + self.block_len;
-
-        let num_blocks = old_len / self.block_len;
-        if self.num_free_blocks <= num_blocks {
-            self.close_block(num_blocks - self.num_free_blocks);
+    fn extend_array(&mut self, helper: &mut BuildHelper) -> Result<()> {
+        if u32::try_from(self.states.len()).unwrap() > u32::MAX - self.block_len {
+            return Err(DaachorseError::automaton_scale("states.len()", u32::MAX));
         }
 
-        for i in old_len..new_len {
-            self.states.push(State::default());
-            self.set_next(i, i + 1);
-            self.set_prev(i, i - 1);
-        }
+        helper.push_block()?;
+        (0..self.block_len).for_each(|_| self.states.push(State::default()));
```

Comment:
Usually, `resize()` is faster than `for_each()`+`push()`.

Reply:
Experiments:
block_size: 1000, extends 100 times, 10000 iters.
* for_each()+push(): 1.69 [sec]
* resize(): 0.55 [sec]

Reply:
Yeah. But, I use `push()` instead of `resize()` here to simply employ an internal reallocator of `Vec`.

Reply:
Oh, I was mistaken that `resize(n)` exactly reserves the capacity of `n` without using an internal reallocator. (i.e., growing in amortized constant-time is not guaranteed.)
I confirmed that `resize` also reserves the capacity as well as `push` at hand.

Reply:
Modified. Thanks. https://github.com/daac-tools/daachorse/pull/49/commits/cd6e18fa9c2dced8829fb4e3452e02f8e3101a92